# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/frontend/BytecodeCompiler.cpp
# Commit: 2ac97be10da5
# Full Hash: 2ac97be10da5628e6ad111f2f363a9dcb5f27847
# Author: Tooru Fujisawa <arai_a@mac.com>
# Date: 2024-12-05 09:58:04
# Regressor Bug: 1933681
# File Overlap Count: 1
# Description:
#   Bug 1933681 - Part 5: Move cached case out of CompileLazyFunctionToStencilMaybeInstantiate. r=nbp
#   
#   Handle the cached case before accessing the script source units, and also
#   make the concurrent delazification case check the cached one, so that
#   the on-demand delazification is reflected to the concurrent delazification.
# ==============================================================================

diff -r 2ae31596c6a9 -r 2ac97be10da5 js/src/frontend/BytecodeCompiler.cpp
--- a/js/src/frontend/BytecodeCompiler.cpp	Wed Dec 04 23:44:34 2024 +0000
+++ b/js/src/frontend/BytecodeCompiler.cpp	Wed Dec 04 23:44:34 2024 +0000
@@ -1436,26 +1436,6 @@
   MOZ_ASSERT(input.source);
 
   AutoAssertReportedException assertException(maybeCx, fc);
-  if (input.options.consumeDelazificationCache()) {
-    const CompilationStencil* cached =
-        stencils->getDelazificationFor(input.extent());
-    if (cached) {
-      if (output.is<RefPtr<CompilationStencil>>()) {
-        output.as<RefPtr<CompilationStencil>>() =
-            const_cast<CompilationStencil*>(cached);
-        assertException.reset();
-        return true;
-      }
-
-      MOZ_ASSERT(maybeCx);
-
-      if (!InstantiateLazyFunction(maybeCx, input, *cached, output)) {
-        return false;
-      }
-      assertException.reset();
-      return true;
-    }
-  }
 
   InheritThis inheritThis =
       input.functionFlags().isArrow() ? InheritThis::Yes : InheritThis::No;
@@ -1597,6 +1577,31 @@
 
   AutoIncrementalTimer timer(cx->realm()->timers.delazificationTime);
 
+  JS::CompileOptions options(cx);
+  options.setMutedErrors(lazy->mutedErrors())
+      .setFileAndLine(lazy->filename(), lazy->lineno())
+      .setColumn(JS::ColumnNumberOneOrigin(lazy->column()))
+      .setScriptSourceOffset(lazy->sourceStart())
+      .setNoScriptRval(false)
+      .setSelfHostingMode(false)
+      .setEagerDelazificationStrategy(lazy->delazificationMode());
+
+  Rooted<CompilationInput> input(cx, CompilationInput(options));
+  input.get().initFromLazy(cx, lazy, ss);
+
+  RefPtr<InitialStencilAndDelazifications> stencils =
+      lazy->sourceObject()->maybeGetStencils();
+
+  if (stencils && input.get().options.consumeDelazificationCache()) {
+    const CompilationStencil* cached =
+        stencils->getDelazificationFor(input.get().extent());
+    if (cached) {
+      CompilationGCOutput* unusedGcOutput = nullptr;
+      BytecodeCompilerOutput output(unusedGcOutput);
+      return InstantiateLazyFunction(cx, input.get(), *cached, output);
+    }
+  }
+
   size_t sourceStart = lazy->sourceStart();
   size_t sourceLength = lazy->sourceEnd() - lazy->sourceStart();
 
@@ -1613,21 +1618,6 @@
     return false;
   }
 
-  JS::CompileOptions options(cx);
-  options.setMutedErrors(lazy->mutedErrors())
-      .setFileAndLine(lazy->filename(), lazy->lineno())
-      .setColumn(JS::ColumnNumberOneOrigin(lazy->column()))
-      .setScriptSourceOffset(lazy->sourceStart())
-      .setNoScriptRval(false)
-      .setSelfHostingMode(false)
-      .setEagerDelazificationStrategy(lazy->delazificationMode());
-
-  Rooted<CompilationInput> input(cx, CompilationInput(options));
-  input.get().initFromLazy(cx, lazy, ss);
-
-  RefPtr<InitialStencilAndDelazifications> stencils =
-      lazy->sourceObject()->maybeGetStencils();
-
   CompilationGCOutput* unusedGcOutput = nullptr;
   BytecodeCompilerOutput output(unusedGcOutput);
   return CompileLazyFunctionToStencilMaybeInstantiate(
@@ -1669,6 +1659,15 @@
     ScopeBindingCache* scopeCache, CompilationStencil& context,
     ScriptIndex scriptIndex, InitialStencilAndDelazifications* stencils,
     DelazifyFailureReason* failureReason) {
+  MOZ_ASSERT(stencils);
+
+  const CompilationStencil* cached = stencils->getDelazificationAt(scriptIndex);
+  if (cached) {
+    RefPtr<CompilationStencil> stencil =
+        const_cast<CompilationStencil*>(cached);
+    return stencil.forget();
+  }
+
   ScriptStencilRef script{context, scriptIndex};
   const ScriptStencilExtra& extra = script.scriptExtra();
 
